state-chain/pallets/cf-ingress-egress/src/lib.rs

```diff
@@ -1251,6 +1327,26 @@ impl<T: Config<I>, I: 'static> IngressSink for Pallet<T, I> {
 }
 
 impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	fn mark_transaction_as_tainted_inner(
+		origin: OriginFor<T>,
+		tx_id: <T::TargetChain as Chain>::DepositDetails,
+	) -> DispatchResult {
+		let account_id = ensure_signed(origin)?;
+		let is_broker = T::AccountRoleRegistry::has_account_role(&account_id, AccountRole::Broker);
+		let is_lp =
+			T::AccountRoleRegistry::has_account_role(&account_id, AccountRole::LiquidityProvider);
+		ensure!(is_broker || is_lp, BadOrigin);
+		let tainted_transaction = TaintedTransactionDetails {
+			broker: account_id.clone(),
+			refund_address: None,
+			deposit_witness: None,
+			expires_at: <frame_system::Pallet<T>>::block_number()
+				.saturating_add(BlockNumberFor::<T>::from(TAINTED_TX_EXPIRATION_BLOCKS)),
+			marked_for_refund: false,
```

Comment:
I think it would be clearer/safer/more efficient if we split this into two storage items: 

```rust
// in a map of AccountId -> TxId -> TaintedTransactionDetails
struct TaintedTransactionDetails {
	// I don't think we need this, it's already in the key of the map:
	// reporter: AccountId,
	expires_at: BlockNumber,
}

// in a value of Vec<RejectionDetails>
struct RejectionDetails {
	refund_address: Address,
	deposit_witness: DepositWitness,
}
```

This avoids having to deal with the `Option`s and filter against `marked_for_refund` etc.

Reply:
I don't think we need DepositWitness tbh. Definitely not for BTC (tx_id is enough here) and for ETH we need other stuff I (refund_address, asset, amount). Because we need to pass this information to the API call, I think we're going to end up to define a new type for that anyway. I would go for now with something static that works for ETH/BTC and extend the type as we go or make generic for the chain.